File: examples/part2/move_semantics.cpp
#include <vector>
#include <string>
#include <iostream>

class Resource {
public:
    std::vector<int> data;
    std::string name;
    
    Resource(size_t size, const std::string& n) 
        : data(size, 42), name(n) {
        std::cout << "Created " << name << " with " << size << " elements\n";
    }
    
    // Copy constructor - expensive!
    Resource(const Resource& other) 
        : data(other.data), name(other.name) {
        std::cout << "COPIED " << name << " (expensive!)\n";
    }
    
    // Move constructor - efficient, but optional
    Resource(Resource&& other) noexcept
        : data(std::move(other.data)), name(std::move(other.name)) {
        std::cout << "MOVED " << name << " (efficient!)\n";
    }
};

void cpp_move_semantics() {
    Resource r1(1000000, "BigResource");
    
    // Accidental copy - compiles but slow!
    Resource r2 = r1;  // Copy constructor called
    
    // Explicit move - fast but easy to forget
    Resource r3 = std::move(r1);  // Move constructor called
    
    // r1 is now in "valid but unspecified state" - dangerous!
    // std::cout << r1.name;  // Might work, might not
    
    std::vector<Resource> vec;
    Resource temp(500000, "TempResource");
    
    // Without std::move, this copies!
    vec.push_back(temp);  // COPY - programmer forgot std::move
    
    // With std::move, this moves
    vec.push_back(std::move(temp));  // MOVE - but temp is now unusable
}

int main() {
  cpp_move_semantics();
  return 0;
}

File: examples/part2/move_semantics.rs
#![allow(dead_code)]

#[derive(Debug)]
struct Resource {
    data: Vec<i32>,
    name: String,
}

impl Resource {
    fn new(size: usize, name: &str) -> Self {
        println!("Created {} with {} elements", name, size);
        Resource {
            data: vec![42; size],
            name: name.to_string(),
        }
    }
}

fn rust_move_semantics() {
    let r1 = Resource::new(1000000, "BigResource");

    // Move by default - always efficient!
    let r2 = r1; // r1 is moved, not copied

    // println!("{:?}", r1);  // Compile error: value used after move
    println!("{:?}", r2.name); // Only r2 is valid

    let mut vec = Vec::new();
    let temp = Resource::new(500000, "TempResource");

    // Move by default - no std::move needed!
    vec.push(temp); // temp is moved into vec

    // println!("{:?}", temp);  // Compile error: value used after move

    // If you actually want to copy, you must be explicit
    #[derive(Debug, Clone)] // Must implement Clone trait
    struct CopyableResource {
        small_data: i32,
    }

    let copyable = CopyableResource { small_data: 42 };
    let copied = copyable.clone(); // Explicit copy
    println!("Original: {:?}, Copy: {:?}", copyable, copied); // Both valid
}

fn main() -> () {
    rust_move_semantics()
}
